stdlib/public/runtime/HeapObject.cpp

```diff
@@ -66,6 +66,10 @@ static inline bool isValidPointerForNativeRetain(const void *p) {
   // arm64_32 is special since it has 32-bit pointers but __arm64__ is true.
   // Catch it early since __POINTER_WIDTH__ is generally non-portable.
   return p != nullptr;
+#elif defined(__ANDROID__) && defined(__aarch64__)
+  // Check the top of the second byte instead, since Android AArch64 reserves
+  // the top byte for its own pointer tagging since Android 11.
+  return (intptr_t)p << 8 > 0;
```

Comment:
Same, someone else should be commenting on the impact here and whether we could clean up how we define our ABI some.

Reply:
@jckarter, you added the negative check below in #7732: let me know if this is a good way to do this, given that Android AArch64 reserves the top byte and I moved the immortal bit to the top of the second byte.

Reply:
As you've written it, this is undefined behavior if tag bytes are shifted out because of signed overflow. However, the goal on other 64-bit platforms here is to reserve some address space for "free" for small closure contexts and such, since checking for `x == 0` and `x < 0` are both one CPU instruction, but I don't think you can do the comparison you'd need here without putting another instruction on the `swift_retain` hot path. We don't currently take advantage of this optimization, so you could do `p != nullptr` like on 32-bit platforms for now.

Reply:
@buttaface ?

Reply:
> this is undefined behavior if tag bytes are shifted out because of signed overflow

I'll look into it, and try your alternative.

Reply:
> `p != nullptr`

This doesn't work, presumably because it never checks for the immortal bit, ie what you check on 64-bit platforms with `p < 0`.

I looked into why the left shift is considered undefined behavior, and as you noted it's because of overflow, but that only matters if you care about doubling the number by shifting, which we don't care about here.

The aim here is simply to check if the ninth bit is set and if the pointer is non-zero if that immortal bit isn't set, ie the same checks as the 64-bit platforms below but simply with an immortal bit located a byte down, so I think this left shift I used should work fine.

Reply:
@buttaface can you address the undefined behavior aspect?

Reply:
Unfortunately, that's not what undefined behavior means. I'm not concerned UB might produce the wrong results, because that would mean we actually got lucky.

You can use an unsigned representation to make it defined or otherwise work around the issue

Reply:
OK, I probably don't understand fully what UB means in the context of this C++ then. My understanding is that left shifts are defined on signed integers too, but they may overflow if the input number to be shifted gets too large, just like with unsigned integers.

Since in this case, I know exactly what the input is, ie a tag in the top byte that I _want_ shifted out, and an immortal bit in the ninth bit that I then want to run the same `> 0` check on, I don't see this as UB. Changing this to unsigned would defeat the purpose, which is to use the same `> 0` check as used for other 64-bit platforms a couple lines below.

Reply:
I looked into this more and apparently [there is a nonsense declaration in some recent C++ versions](https://en.cppreference.com/w/cpp/language/operator_arithmetic) that `For negative a, the behavior of a << b is undefined`. Since this can only be enforced if `a` is known at compile-time or by inserting a runtime check to see if `a` is negative before shifting, [Regehr notes it is never enforced as it would break a ton of code](https://blog.regehr.org/archives/738).

My intent here was to keep this quick check for this commonly invoked function and the shifted version compiles down to the following tight three instructions, after I disabled the mandatory inlining:
```
       0: 08 dc 78 d3  	lsl	x8, x0, #8
       4: 1f 01 00 f1  	cmp	x8, #0                  // =0
       8: e0 d7 9f 1a  	cset	w0, gt
       c: c0 03 5f d6  	ret
```
I then wrote the equivalent by masking and comparing twice, with no shifts, pushing it up to five instructions:
```
       0: 1f 00 49 f2  	tst	x0, #0x80000000000000
       4: e8 17 9f 1a  	cset	w8, eq
       8: 1f dc 40 f2  	tst	x0, #0xffffffffffffff
       c: e9 07 9f 1a  	cset	w9, ne
      10: 00 01 09 0a  	and	w0, w8, w9
      14: c0 03 5f d6  	ret
```
I'm skeptical that any C++ compiler would ever dare break this commonly used idiom, so I think we should just use the shifted version, which I've commented out in this pull for now. If you don't want to do that, I'm fine with the longer mask and compares version instead.

@drodriguez, you may want to chime on this and the whole pull.

Reply:
Not my area of expertise, but I agree with Michael and Joe that using undefined behaviour because it works for your use case today is not the way to go.

I was trying  to understand what we were trying, and it seems that you want to move the bit 56 into the sign bit of the 64 bit register (assuming two's complement), but I think you can also go the other way:

```
return (int32_t)((intptr_t)p >> 16) > 0;
```

This end up with 3 instructions in Godbolt with -O3, which put it on par to the solution that relies on UB:

```
        lsr     x8, x0, #16
        cmp     w8, #0                          // =0
        cset    w0, gt
        ret
```

Double-check that my idea is solid and return the same results, and wait to see if anyone else can find the hole in my proposal.

PD: I skimmed the rest of the proposal, and I think there's some cases where you write comments about the 24th bit, and I think you meant the 56th bit.

Reply:
> it seems that you want to move the bit 56 into the sign bit of the 64 bit register

That's part of it: as Joe noted above, we need to know if the immortal bit is set, `x < 0` normally, and that the pointer is non-null, `x != 0`. Your alternate formulation is throwing away information for the second check and will only work if you're sure that no pointer will only be contained in the bottom 16 bits.

My most recent mask and compares version in this pull checks both completely.

> there's some cases where you write comments about the 24th bit, and I think you meant the 56th bit.

I think you're referring to the metadata changes alone. There, the compiler places two 32-bit words into a 64-bit pointer in the uninitialized cache: one containing an Int32 with the string's size and the other with a 32-bit relative address to the mangled string. So in that case, my comments about the 24th bit only refers to that first 32-bit word with the size, since that's the only one that's checked, even though we check it with a 64-bit mask to the entire pointer.

Reply:
> Your alternate formulation is throwing away information for the second check

I was trying to reproduce your initial code (`(intptr_t)p << 8 > 0`) but instead of moving `<<`, moving `>>` and leaving the 56th bit on the 32th place and using a 32 bit comparison for positive (maybe the offset should be 24 and not 16 in my code?). Was your initial code also not having into account the second check?

> the compiler places two 32-bit words

That makes sense. As I said, I just skimmed, and the discussion talking about 24th was unexpected.

Reply:
> maybe the offset should be 24 and not 16 in my code?

Yep, should be 24.

> Was your initial code also not having into account the second check?

Let me back up a bit and explain what we're doing here. This inlined function `isValidPointerForNativeRetain()` is used throughout the ARC functions in this file to determine if ARC should run or not. It does so by checking if the pointer has the immortal bit set, which on 64-bit platforms is the top "sign" bit as can be seen elsewhere in this pull, or if it is a null pointer, and it ignores the pointer in both cases.

On Android AArch64 with memory tagging, the top byte of these pointers is reserved for the OS, so Swift's immortal bit can no longer be in the top bit, and I move it one byte down to the ninth bit in this pull. The shifted version of this check tries to preserve that same quick `p > 0` comparison by shifting the immortal bit back into the top sign bit and a shifted null pointer is still a null pointer.

What I have in this pull now by masking and comparing is the explicit version of the same check:
```
((0x0080000000000000ULL & (uintptr_t)p) == 0) && ((0x00FFFFFFFFFFFFFFULL & (uintptr_t)p) > 0)
```
I think we should use the quicker shifted version, but I'm fine with this instead.

Reply:
If you want to use the shifted version, then you can use unsigned rather than signed types. But I don't think any of this is necessary—it's sufficient and cheaper to check for `!= 0`. We do not currently use the negative bit optimization for anything on existing 64-bit platforms, and when we do, we would need to disable or modify those optimizations for MTE-enabled Android anyway, since there would be fewer bits available for storing data than on other platforms.

Reply:
> If you want to use the shifted version, then you can use unsigned rather than signed types.

I'll try that: I'll see if clang accepts shifting it as an unsigned type, then casting the result to a signed type for the comparison.

> But I don't think any of this is necessary—it's sufficient and cheaper to check for != 0. We do not currently use the negative bit optimization for anything on existing 64-bit platforms, and when we do, we would need to disable or modify those optimizations for MTE-enabled Android anyway, since there would be fewer bits available for storing data than on other platforms.

You're saying the top bit is currently never intentionally set for pointers as an immortal bit on 64-bit platforms? Or that we should now ignore the immortal bit when set on Android AArch64? Because when I tried your `p != nullptr` alternative earlier, several stdlib tests started failing, whereas my two alternatives here that check the immortal bit keep them passing.

I can investigate further why that is, if you believe only checking for null pointers should suffice.

Reply:
Hm, which tests are those that fail?

Reply:
I ran a little experiment to see if I can reproduce the stdlib test failures on linux x86_64 and I can. I setup a Ubuntu 20.04 VPS and downloaded the latest Mar. 30 trunk snapshot build and source, before running the following command to build the standalone stdlib and run its tests:
```
./swift/utils/build-script -RA --build-swift-tools=0 --skip-build-cmark --build-llvm=0
--native-clang-tools-path=/home/butta/swift-DEVELOPMENT-SNAPSHOT-2022-03-30-a-ubuntu20.04/usr/bin
--native-swift-tools-path=/home/butta/swift-DEVELOPMENT-SNAPSHOT-2022-03-30-a-ubuntu20.04/usr/bin
-j9 -T --lit-args=--filter=stdlib/ --skip-test-early-swift-driver
```
That runs about 530 stdlib tests and only 5 fail, because I don't have needed utilities like llvm-nm installed.

Then I changed the line below from `return (intptr_t)p > 0` to `return p != nullptr` and reran the above build command, only to see 350 tests failing now. Looking at the results, I see several segfaults like this, ie many test binaries segfault if we only check for the null pointer:
```
/home/butta/build/Ninja-Release/swift-linux-x86_64/validation-test-linux-x86_64/stdlib/Output/Bitset.swift.script:
line 1: 356022 Segmentation fault      (core dumped)
/usr/bin/env LD_LIBRARY_PATH='/home/butta/build/Ninja-Release/swift-linux-x86_64/lib/swift/linux'
/home/butta/build/Ninja-Release/swift-linux-x86_64/validation-test-linux-x86_64/stdlib/Output/Bitset.swift.tmp/a.out
```
Around the same number of stdlib tests failed on Android AArch64 when I only checked for the null pointer a couple weeks ago as you suggested, which you can reproduce on linux if you like using the above steps: let me know if it is unexpected and if we should investigate further.

Reply:
I've updated this line to shift again with `return (intptr_t)((uintptr_t)p << 8) > 0`, which maintains the same shifting three-instruction assembly output above and hopefully threads your UB needle too.